man/uialert.1
.\"-
.\" Copyright (c) 2020-2021 ProcursusTeam
.\" SPDX-License-Identifier: BSD-4-Clause
.\"
.Dd October 15, 2021
.Dt UIALERT 1
.Os
.Sh NAME
.Nm uialert
.Nd Display an alert
.Sh SYNOPSIS
.Nm
.Op Fl b Ar body
.Op Fl i Ar prompt
.Op Fl j
.Op Fl p Ar primary
.Op Fl -priority Ar number
.Op Fl s Ar second
.Op Fl -secure Ar prompt
.Op Fl t Ar third
.Op Fl -timeout Ar number
.Ar title
.Sh DESCRIPTION
Display a native alert in the GUI from the command line.
The title will be set to
.Ar title .
It will print the response to each input and the index of the selected button to
.Ar stdout .
.Ss OUTPUT
.Bl -tag -width 4n -offset indent -compact
.It Sy 0
The primary button was pressed.
.It Sy 1
The secondary button was pressed.
.It Sy 2
The tertiary button was pressed.
.It Sy 3
The alert timed out or was closed without a response.
.El
.Sh OPTIONS
.Bl -tag -width indent
.It Fl b , -body Ar body
Set the alert's body text to
.Ar body .
.It Fl i , -inpute Ar prompt
Add a text box with a prompt of
.Ar prompt .
This can be specified multiple times to create more inputs.
.It Fl j , -json
Print the output in JSON.
.It Fl p , -primary Ar primary
Set the default button's text to be
.Ar primary ,
instead of the default of
.Qq OK .
.It Fl -priority Ar number
Set the priority of the alert to be
.Ar number .
.Ar number
must be between 0 and 3, inclusive.
If not specified the priority will be 1.
This will only affect the appearence of the alert on
.Os macOS.
.It Fl s , -secondary Ar secondary
Set the second button's text to be
.Ar secondary ,
if unset, this button will not show.
.It Fl -secure Ar prompt
This is identical to
.Fl i ,
except the inputed text will be treated as a password.
.It Fl t , -tertiary Ar third
Set the third button's text to be
.Ar third ,
if unset, this button will not show.
.It Fl -timeout Ar number
After
.Ar number
seconds, dismiss the alert.
.El
.Sh EXAMPLES
Create an alert with one button that says
.Qq OK
.Dl "uialert -b \*qThis is the body\*q \*qAn Interesting Title\*q"
.Pp
Create an alert with two buttons, where the default button says
.Qq Cancel .
.Dl "uialert -p \*qCancel\*q -s \*qContinue Anyways\*q -b \*qContinue?\*q \*qThis Action Will Delete Data\*q"
.Sh HISTORY
The
.Nm
utility first appeared in uikittools on Feburary 26, 2008, written by
.An Jay Qo Saurik Qc Freeman .
Later, it was reimplemented by
.An Cameron Katri
for uikittools-ng on October 15, 2021 with completely different options and calling.


uialert.m
#include <CoreFoundation/CoreFoundation.h>
#include <Foundation/Foundation.h>
#include <err.h>
#include <getopt.h>
#include <stdio.h>
#include <stdlib.h>

#ifndef NO_NLS
#	include <libintl.h>
#	define _(a) gettext(a)
#	define PACKAGE "uikittools-ng"
#else
#	define _(a) a
#endif

#ifndef LOCALEDIR
#	define LOCALEDIR "/usr/share/locale"
#endif

// clang-format off
void usage() {
	printf(_("Usage: %s [-b body] [-i prompt] [-j] [-p primary] [--priority number] [-s second] [--secure prompt] [-t third] [--timeout number] title\n\
Copyright (C) 2021, Procursus Team. All Rights Reserved.\n\n\
Display an alert\n\n"), getprogname());

	printf(_("  -b, --body <text>        Text for alert body\n\
  -p, --primary <text>     Default button text instead of \"OK\"\n\
      --priority 0-3       Alert priority\n\
                           This will change the icon on macOS\n\
  -s, --secondary <text>   Second button text\n\
  -t, --tertiary <text>    Third button text\n\
  -i, --input <text>       Ask for text input\n\
  -j, --json               Print the output in JSON\n\
      --secure <text>      Show text field as secure\n\
      --timeout <num>      Number of seconds to wait before exiting\n\n"));

	printf(_("Output:\n\
  0 - primary button\n\
  1 - secondary button\n\
  2 - tertiary button\n\
  3 - timeout/cancel\n\n\
Contact the Procursus Team for support.\n"));
	exit(1);
}
// clang-format on

enum {
	OPT_PRIORITY = CHAR_MAX + 1,
	OPT_TIMEOUT,
	OPT_SECURE,
};

int main(int argc, char **argv) {
#ifndef NO_NLS
	setlocale(LC_ALL, "");
	bindtextdomain(PACKAGE, LOCALEDIR);
	textdomain(PACKAGE);
#endif

	CFOptionFlags flags;
	double timeout = 0, index = 0;
	int priority = kCFUserNotificationNoteAlertLevel;
	char *message = NULL;
	char *defaultButton = NULL;
	char *alternativeButton = NULL;
	char *otherButton = NULL;
	int ch;
	const char *errstr;
	bool json = false;
	NSError *error;

	NSMutableDictionary *alert = [[NSMutableDictionary alloc] init];
	NSMutableArray *inputs= [[NSMutableArray alloc] init];

	// clang-format off
	static struct option longopts[] = {
		{"body", required_argument, NULL, 'b'},
		{"primary", required_argument, NULL, 'p'},
		{"priority", required_argument, NULL, OPT_PRIORITY},
		{"secondary", required_argument, NULL, 's'},
		{"tertiary", required_argument, NULL, 't'},
		{"timeout", required_argument, NULL, OPT_TIMEOUT},
		{"input", required_argument, NULL, 'i'},
		{"json", no_argument, NULL, 'j'},
		{"secure", required_argument, NULL, OPT_SECURE},
		{NULL, 0, NULL, 0}};
	// clang-format on

	while ((ch = getopt_long(argc, argv, "b:p:s:t:i:j", longopts, NULL)) != -1) {
		switch (ch) {
			case 'b':
				[alert addEntriesFromDictionary:@{
					(__bridge NSString *)kCFUserNotificationAlertMessageKey:
					[NSString stringWithUTF8String:optarg]}];
				break;
			case 'p':
				[alert addEntriesFromDictionary:@{
					(__bridge NSString *)kCFUserNotificationDefaultButtonTitleKey:
					[NSString stringWithUTF8String:optarg]}];
				break;
			case 's':
				[alert addEntriesFromDictionary:@{
					(__bridge NSString *)kCFUserNotificationAlternateButtonTitleKey:
					[NSString stringWithUTF8String:optarg]}];
				break;
			case 't':
				[alert addEntriesFromDictionary:@{
					(__bridge NSString *)kCFUserNotificationOtherButtonTitleKey:
					[NSString stringWithUTF8String:optarg]}];
				break;
			case OPT_SECURE:
				flags |=
					CFUserNotificationSecureTextField(inputs.count);
			case 'i':
				[inputs addObject:[NSString stringWithUTF8String:optarg]];
				break;
			case 'j':
				json = true;
				break;
			case OPT_PRIORITY:
				switch (strtonum(optarg, 0, 3, &errstr)) {
					case 0:
						if (errstr != NULL)
							err(1, "%s", optarg);
						priority = kCFUserNotificationPlainAlertLevel;
						break;
					case 1:
						priority = kCFUserNotificationNoteAlertLevel;
						break;
					case 2:
						priority = kCFUserNotificationCautionAlertLevel;
						break;
					case 3:
						priority = kCFUserNotificationStopAlertLevel;
						break;
				}
				break;
			case OPT_TIMEOUT:
				timeout = strtonum(optarg, 0, INT_MAX, &errstr);
				if (errstr != NULL)
					err(1, "%s", optarg);
				break;
			default:
				usage();
		}
	}
	argc -= optind;
	argv += optind;

	flags |= priority;

	if (argc != 1) usage();

	if (argv[0] == NULL) usage();

	[alert addEntriesFromDictionary:@{
		(__bridge NSString *)kCFUserNotificationAlertHeaderKey:
		[NSString stringWithUTF8String:argv[0]]}];

	if (inputs.count > 0)
		[alert addEntriesFromDictionary:@{
			(__bridge NSString *)kCFUserNotificationTextFieldTitlesKey:
			inputs}];

	CFUserNotificationRef notif = CFUserNotificationCreate(
		kCFAllocatorDefault, timeout, flags, NULL, (__bridge CFMutableDictionaryRef)alert);

	CFOptionFlags cfRes;

	CFUserNotificationReceiveResponse(notif, 0, &cfRes);

	NSDictionary *response = (__bridge NSDictionary *)CFUserNotificationGetResponseDictionary(notif);

	NSMutableDictionary *out = [[NSMutableDictionary alloc] init];

	if (cfRes != kCFUserNotificationCancelResponse && inputs.count >= 1) {
		NSObject *input = [response valueForKey:(__bridge NSString *)kCFUserNotificationTextFieldValuesKey];

		if ([input isKindOfClass:NSString.class]) {
			[out addEntriesFromDictionary:@{
				@"inputs" : @[(NSString *)input]
			}];
		} else if ([input isKindOfClass:NSArray.class]) {
			[out addEntriesFromDictionary:@{
				@"inputs" : (NSArray *)input
			}];
		}
	}

	[out addEntriesFromDictionary:@{
		@"button" : [NSNumber numberWithUnsignedLong:cfRes]
	}];

	if (json) {
		NSData *jsonData;

		jsonData = [NSJSONSerialization dataWithJSONObject:out options:0 error:&error];

		if (error)
			errx(1, _("JSON formating failed: %s"), error.localizedDescription.UTF8String);

		printf("%s\n", [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding].UTF8String);
	} else {
		for (NSString *item in [out valueForKey:@"inputs"])
			printf("%s\n", item.UTF8String);

		printf("%lu\n",
			   [(NSNumber *)[out valueForKey:@"button"] unsignedLongValue]);
	}

	CFRelease(notif);
	return 0;
}
